staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto

```diff
@@ -1531,6 +1559,27 @@ message PullImageResponse {
     // Reference to the image in use. For most runtimes, this should be an
     // image ID or digest.
     string image_ref = 1;
+    // Reference to a local path on disk if mounted.
+    // Introduced in the OCI Volume Source KEP: https://kep.k8s.io/4639
+    string mount_ref = 2;
+    // The supported source type if the image got mounted as OCI volume.
+    // The kubelet uses this field to check for availability of the feature.
+    // Introduced in the OCI Volume Source KEP: https://kep.k8s.io/4639
+    OCIVolumeMountSource source = 3;
+}
+
+// OCIVolumeMountSource describes the supported source types for an OCI volume mount.
+// Introduced in the OCI Volume Source KEP: https://kep.k8s.io/4639
+enum OCIVolumeMountSource {
+    // Unsupported is the default source mount type and indicates that the
+    // feature is not available by the runtime.
+    UNSUPPORTED = 0;
```

Comment:
What does the kubelet do in this case if it set mount to true? Why not use RuntimeHandlerFeatures to check for this and fail early?

Reply:
The kubelet will throw an event that the feature is not supported and let the container creation fail on `SyncPod` (when the OCI object pull happens).

IIUC, then the `RuntimeHandlerFeatures` would be checked during `SyncPod` as well, so we could add it to them but would not fail earlier.

Any thoughts on that @haircommander @mikebrow ?

Reply:
we also could fail early and check the features at kubelet startup if the feature gate is on, and reject any pods that request the feature

Reply:
also in https://github.com/kubernetes/enhancements/pull/4728 we seem to have agreed upon a CRI level "runtime features" object which I think this would be a better fit for https://github.com/kubernetes/enhancements/pull/4728

Reply:
> we also could fail early and check the features at kubelet startup if the feature gate is on, and reject any pods that request the feature

I'm not a big fan of that considering that runtimes are swappable (restart and upgrade scenarios included). I assume we have to check at least for the runtime handler once per sync pod iteration.

Reply:
I added the `oci_volume_mounts` bool to `RuntimeHandlerFeatures` in favor of the `UNSUPPORTED` enum type.

Reply:
I still would push for RuntimeFeatures object instead. though, @everpeace was this going to make it for 1.31 ?

Reply:
> was this going to make it for 1.31 ?

Yes, I'd like to have it for 1.31. But, as you know, I'm now waiting for https://github.com/kubernetes/enhancements/pull/4728 got approved by @mrunalp.


By the way, I already opened https://github.com/kubernetes/kubernetes/pull/125470 which introduces `NodeFeatures`(k8s API) and `RuntimeFeatures`(cri). The PR is already ready for review, actually. However, I mark this as WIP because https://github.com/kubernetes/enhancements/pull/4728 has not been merged yet, though.

Reply:
Is a good point that each connection between kubelet and the runtime cri integration could possibly have issues with incompatible feature sets. This will possibly require storing the feature set used for a set of pods on a connection, and then checking for diffs.. any diff indicates a possible issue as the diffs may include removal of a feature or adding a feature that was not supported on running pods. hmm...

Reply:
Is it true that the possibility of configuring incompatible feature sets exists already and wouldn't be introduced with the OCI VolumeMount feature? (I couldn't find a clear example so correct me if I'm wrong, maybe with TopologyManager/NUMA?) If this is the case, would storing feature sets used by pods per connection & checking for diffs be a separate enhancement?

Reply:
> Is it true that the possibility of configuring incompatible feature sets exists already and wouldn't be introduced with the OCI VolumeMount feature? (I couldn't find a clear example so correct me if I'm wrong, maybe with TopologyManager/NUMA?) If this is the case, would storing feature sets used by pods per connection & checking for diffs be a separate enhancement?

yes... For example: not having, or loosing via downgrade/config change, support for a runtime handler type on the node.